src/spark_manager.py

```diff
@@ -0,0 +1,165 @@
+import logging
+import os
+import uuid
+from typing import Dict, List, Optional
+
+import kubernetes as k8s
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class KubeSparkManager:
+    """
+    Manager for user-specific Spark clusters in Kubernetes.
+
+    This class provides methods to create, manage, and destroy Spark clusters
+    for individual JupyterHub users.
+    """
+
+    # Required environment variables
+    REQUIRED_ENV_VARS = {
+        "KUBE_NAMESPACE": "Kubernetes namespace for Spark clusters",
+        "SPARK_IMAGE": "Docker image for Spark master and workers",
+        "POSTGRES_USER": "PostgreSQL username",
+        "POSTGRES_PASSWORD": "PostgreSQL password",
+        "POSTGRES_DB": "PostgreSQL database name",
+        "POSTGRES_URL": "PostgreSQL connection URL"
+    }
+
+    # Default configuration values for cluster settings
+    DEFAULT_WORKER_COUNT = 2
+    DEFAULT_WORKER_CORES = 10
+    DEFAULT_WORKER_MEMORY = "10G"
+    DEFAULT_MASTER_CORES = 10
+    DEFAULT_MASTER_MEMORY = "10G"
+
+    DEFAULT_IMAGE_PULL_POLICY = os.environ.get("SPARK_IMAGE_PULL_POLICY", "IfNotPresent")
+
+    DEFAULT_EXECUTOR_CORES = 2
+    DEFAULT_MAX_CORES_PER_APPLICATION = 10
+    DEFAULT_MAX_EXECUTORS = 5
+
+    DEFAULT_MASTER_PORT = 7077
+    DEFAULT_MASTER_WEBUI_PORT = 8090
+    DEFAULT_WORKER_WEBUI_PORT = 8081
+
+    @classmethod
+    def validate_environment(cls) -> Dict[str, str]:
+        """
+        Validate that all required environment variables are set.
+
+        Returns:
+            Dict[str, str]: Dictionary of validated environment variables
+        """
+        missing_vars = []
+        env_values = {}
+
+        for var, description in cls.REQUIRED_ENV_VARS.items():
+            value = os.environ.get(var)
+            if not value or not value.strip():
+                missing_vars.append(f"{var} ({description})")
+            env_values[var] = value
+
+        if missing_vars:
+            raise ValueError(
+                "Missing required environment variables:\n" +
+                "\n".join(f"- {var}" for var in missing_vars)
+            )
+
+        return env_values
+
+    def __init__(self,
+                 username: str,
+                 namespace: Optional[str] = None,
+                 image: Optional[str] = None,
```

Comment:
Since users won't be calling this function I don't understand the use case for overriding the env

Reply:
👍 yea. Make sense - removed those. It's safer we don't allow users to over write the images.